node/src/exchange/server.rs

```diff
@@ -1,52 +1,382 @@
+            }
+            header_request::Data::Hash(hash) => {
+                self.handle_request_by_hash(response_channel, hash);
+            }
+        };
     }
 
-    #[instrument(level = "trace", skip(self))]
     pub(super) fn on_response_sent(&mut self, peer: PeerId, request_id: RequestId) {
-        // TODO
+        trace!("response_sent; request_id: {request_id}, peer: {peer}");
     }
 
-    #[instrument(level = "trace", skip(self))]
     pub(super) fn on_failure(
         &mut self,
         peer: PeerId,
         request_id: RequestId,
         error: InboundFailure,
     ) {
-        // TODO
+        // TODO: cancel job if libp2p already failed it?
+        trace!("on_failure; request_id: {request_id}, peer: {peer}, error: {error:?}");
+    }
+
+    pub fn poll(&mut self, cx: &mut Context<'_>, sender: &mut R) -> Poll<()>
+    where
+        <R as ResponseSender>::Channel: std::fmt::Debug,
+    {
+        let pol = self.rx.poll_recv(cx);
+        if let Poll::Ready(Some((channel, response))) = pol {
+            sender.send_response(channel, response);
+            return Poll::Ready(());
+        }
+
+        Poll::Pending
+    }
+
+    fn handle_request_current_head(&mut self, channel: R::Channel) {
+        let store = self.store.clone();
+        let tx = self.tx.clone();
+
+        spawn(async move {
+            let response = store
+                .get_head()
+                .await
+                .map(|head| head.to_header_response())
+                .unwrap_or_else(|_| HeaderResponse::not_found());
+
+            let _ = tx.send((channel, vec![response])).await;
+        });
+    }
+
+    fn handle_request_by_hash(&mut self, channel: R::Channel, hash: Vec<u8>) {
+        let store = self.store.clone();
+        let tx = self.tx.clone();
+        spawn(async move {
+            let response = if let Ok(hash) = Hash::from_bytes(Algorithm::Sha256, &hash) {
+                store
+                    .get_by_hash(&hash)
+                    .await
+                    .map(|head| head.to_header_response())
+                    .unwrap_or_else(|_| HeaderResponse::not_found())
+            } else {
+                HeaderResponse::invalid()
+            };
+
+            let _ = tx.send((channel, vec![response])).await;
+        });
+    }
+
+    fn handle_request_by_height(&mut self, channel: R::Channel, origin: u64, amount: u64) {
+        let store = self.store.clone();
+        let tx = self.tx.clone();
+
+        spawn(async move {
+            let mut responses = vec![];
+            for i in origin..origin + amount {
+                let response = store
+                    .get_by_height(i)
+                    .await
+                    .map(|head| head.to_header_response())
+                    .unwrap_or_else(|_| HeaderResponse::not_found());
+                responses.push(response);
+            }
+
+            let _ = tx.send((channel, responses)).await;
+        });
```

Comment:
The Go implementation behaves differently: If `origin` does not exist then `NotFound` is send, otherwise if send only up to the height it managed to retrieve. Also I suggest to have a maximum amount, to avoid malicious requests.

Also as optimization we can get at the beginning if `origin` exists.

```rust
        if !store.has_at(origin) {
            self.send_not_found(response);
            return;
        }

        let store = self.store.clone();
        let tx = self.tx.clone();

        spawn(async move {
            let amount = amount.min(512);
            let mut responses = Vec::with_capacity(amount  as usize);

            for i in origin..origin + amount {
                match store.get_by_height(i).await {
                    Ok(header) => responses.push(header.to_header_response()),
                    Err(_) => break,
                }
            }

            if responses.is_empty() {
                responses.push(HeaderResponse::not_found());
            }

            let _ = tx.send((channel, responses)).await;
        });
```

Reply:
Fixed the algorithm. As for the optimisation,`store.has_at` is also async, so I don't think we can skip `spawn`ing, I've added `has_at` check to at least avoid allocation at the expense of one extra check (we'd break the loop on first iteration anyway, so not much of a time save here).

Reply:
hmm. maybe in this case we can do something like:

```rust
            let mut responses = vec![];

            for i in origin..origin + amount {
                match store.get_by_height(i).await {
                    Ok(header) => {
                        if responses.is_empty() {
                            responses.reserve(amount as usize);
                        }
                        responses.push(header.to_header_response());
                    }
                    Err(_) => break,
                }
            }
```

Reply:
Nice! Removed the unneeded double check.